### PR TITLE
Adding support for renaming and cloning datasets

### PR DESCRIPTION
**Pull request by brimoor**
Implements the ability to clone and rename datasets, as per the discussion in https://github.com/voxel51/fiftyone/issues/385.

To achieve this, the following changes were made:
- sample collections for datasets are given permanent, unique names based on the timestamp of when the dataset was created, not the name of the dataset, which can change
- a `Dataset`'s name can be changed via simply `dataset.name = new_name`. This change is saved in the dataset's backing `DatasetDocument`, as expected
- for clarity, the DB collection storing `DatasetDocument`s is renamed to `datasets`

The sample collection structure in the DB is now:

```
datasets                    # collection storing all DatasetDocuments
samples.<dataset-uuid1>     # a dataset's sample collection
samples.<dataset-uuid2>     # another dataset's sample collection
```

where, currently `<uuid>` is set based on the timestamp of when the dataset was created: `YYYY.MM.DD.HH.MM.SS[.MMM]`

**TODO**
- `Sample`s currently reference their parent dataset via the name of their `SampleDocument` class, which is no longer valid (the dataset name is not the same as the sample collection's name). I always thought `Sample`s should store an actual reference to their parent `Dataset`, so this will be a convenient excuse to update the implementation to reflect that
- optimize the implementation of `Dataset.clone()` to avoid DB I/O; the copy should be easy enough to perform at the DB-level

### CODE REVIEW

**Comment by tylerganter** [comment]
`Sample` objects now have reference to their dataset (if any), and `Dataset.clone` is very fast! How fast you might ask? Well when testing with a dummy dataset with 100 samples with 100 detections each, I timed the two approaches:
```
previous clone: 3.374395
Fast clone: 0.071713
```

File: fiftyone/core/odm/sample.py [hunk 1]
```diff
@@ -232,9 +232,9 @@ def __setattr__(self, name, value):
         super().__setattr__(name, value)
 
     @property
-    def dataset_name(self):
-        """The name of the dataset to which this sample belongs, or ``None`` if
-        it has not been added to a dataset.
+    def collection_name(self):
```
Comment (brimoor) [inline fiftyone/core/odm/sample.py:235]: @tylerganter what do you think about making this private? Users shouldn't need to be concerned with the internal-facing name of the sample collection

File: fiftyone/core/odm/sample.py [hunk 2]
```diff
@@ -438,17 +438,16 @@ def delete_field(cls, field_name):
...
+        )
+        dataset_doc.sample_fields = [
+            sf for sf in dataset_doc.sample_fields if sf.name != field_name
         ]
-        dataset._meta.save()
+        dataset_doc.save()
 
     def _get_repr_fields(self):
-        return ("dataset_name",) + super()._get_repr_fields()
+        return ("collection_name",) + super()._get_repr_fields()
```
Comment (brimoor) [inline fiftyone/core/odm/sample.py:450]: @tylerganter can we remove `collection_name` since it is now an internal-only concept?

**Comment by brimoor** [comment]
> @brimoor obviously review these changes yourself before merging

@tylerganter thanks for the help, looks great! I left a couple questions for you

File: fiftyone/core/odm/sample.py (hunk 1 above)
Comment (tylerganter) [reply fiftyone/core/odm/sample.py:235]: Isn't this entire class effectively private? I think making it public helps convey that this is something publicly available for use by `fiftyone.core.sample.Sample`

Comment (brimoor) [reply fiftyone/core/odm/sample.py:235]: Good point. I see that you did what I actually should have requested, which was just that this field is not exposed publicly on `fiftyone.core.sample.Sample`

**Pull request merged**